proposals/html-module-spec-changes.md

```diff
@@ -52,65 +71,89 @@ Questions/corrections/feedback are welcome!  I've left TODOs in several places w
+  - To create an HTML module script, given a JavaScript string *source*, an environment settings object *settings*, a URL *baseURL*, and some script fetch options *options*:
+  - 1\.	Let *htmlModuleScript* be a new HTML module script that this algorithm will subsequently initialize.
+  - 2\.	Set *htmlModuleScript*'s settings object to *settings*.
+  - 3\.	Set *htmlModuleScript*'s *base URL* to *baseURL*.
+  - 4\.	Set *htmlModuleScript*'s fetch options to *options*.
+  - 5\. Set *htmlModuleScript*'s *parse error* and *error to rethrow* to null.
+  - 6\. Let *result* be ParseHTMLModule(*source*, *settings's* Realm, *htmlModuleScript*). Note: Passing *htmlModuleScript* as the last parameter here ensures *result*.[[HostDefined]] will be *htmlModuleScript*.  See below bullet for ParseHTMLModule definition.
+  - 7\. For each ScriptEntry *required* of *result*.[[RequestedModules]], such that *required*[[InlineModuleRecord]] == null:
+    - a\.	Let *url* be the result of resolving a module specifier given *htmlModuleScript*'s *base URL* and *required*[[ExternalScriptURL]].
     - b\.	If *url* is failure, then:
       - i\.	Let *error* be a new TypeError exception.
-      - ii\.	Set *htmlModule’s* *parse error* to *error*.
-      - iii\.	Return *htmlModule*.  Note: This step is essentially validating all of the requested module specifiers. We treat a module with unresolvable module specifiers the same as one that cannot be parsed; in both cases, a syntactic issue makes it impossible to ever contemplate instantiating the module later.
-  - 8\.   Set *htmlModule’s* *record* to *result*.
-  - 9\.   Return *htmlModule*.
-- Define ParseHTMLModule(*source*, *realm*, *htmlModule*) as roughly the following.  TODO Given that we define  HTML Module Record in ES6, should this function be defined over there?  We’re using the HTML5 parser though...
-  - 1\. Let *record* be a new HTML Module record that this algorithm will subsequently initialize. 
-  - 2\.	Run the HTML5 parser on source to obtain the result *document*.
-  - 3\.	Set *record*.[[HostDefined]] = *htmlModule*.
-  - 4\.	For each HTMLScriptElement *script* in *document*:
-    - a\.	Let *se* be a new ScriptEntry record (see definition in ES6 changes above).
+      - ii\.	Set *htmlModuleScript*'s *parse error* to *error*.
+      - iii\.	Return *htmlModuleScript*.  Note: This step is essentially validating all of the requested module specifiers. We treat a module with unresolvable module specifiers the same as one that cannot be parsed; in both cases, a syntactic issue makes it impossible to ever contemplate instantiating the module later.
+  - 8\. Set *htmlModuleScript*'s *record* to *result*.
+  - 9\. Return *htmlModuleScript*.
+- Introduce a new algorithm ParseHTMLModule(*source*, *realm*, *htmlModuleScript*) as the following.
```

Comment:
I think this needs to be more detailed. In particular what I think we want to do is to run the parser in a mode similar to XMLHttpRequest, whereby nothing is executed or fetched. And then once done we walk the resulting tree and execute scripts provided they meet the requirements.

It also seems reasonable to me to not require the `type` attribute as this is a different processing model so we can also offer developer conveniences.

Reply:
Agreed that this needs to be fleshed out more.  Script execution should already be [disabled](https://html.spec.whatwg.org/#concept-n-noscript) because the module doc doesn't have a browsing context, but the case for fetches isn't so clear.  Most things (`<img>`, `<iframe>`, etc) won't kick off fetches until moved to the main document, but it's possible something I'm not thinking of could slip through and perhaps we should head it off by enforcing a special parser mode.  I'd prefer not to block all the other changes on this though so I added a TODO to investigate/deliberate this further.

The "walking through the resulting tree and executing scripts" process is already handled; see the "For each HTMLScriptElement *script* in *document*" part where we feed the scripts into the ES Modules infra (as the ScriptEntries of the HTML Module Record), so they'll be executed as part of the module graph's Evaluation phase.

Our original proposal had it such that we'd coerce non-module type scripts to module-type, but that got pushback from a few differnent folks that it would be confusing to devs and we should just make non-module scripts an error.  I'm not totally sold on either direction; do you think it's worth opening a new issue thread to discuss this specifically?

Reply:
Sounds good to me. I don't think it'll be confusing long term as HTML modules are quite distinct from HTML documents in many respects.

Reply:
(And thanks for explaining the way script execution happens. That makes sense I think.)

Reply:
I belatedly opened an issue [here](https://github.com/WICG/html-modules/issues/2) for the question of whether to convert classic scripts to module scripts or to throw an error.

It's in the new WICG/html-modules repo for now but I'll move it if we end up not using that.

Reply:
I think we'd make faster progress here. There's 290 people watching this repository. There's 6 people watching that.

Reply:
Yup, we moved it to https://github.com/w3c/webcomponents/issues/798.